### PR TITLE
Banking-app defines its own constitution

### PR DESCRIPTION
**Pull request by ross-p-smith**
Quite a few changes here for what should have been a simple change.

# Why the change
The Banking Sample should define its own constitution rather than copying files over the top of /opt/ccf/bin. We need a change to ccf for this which wont be until v3 [https://github.com/microsoft/CCF/pull/4628](https://github.com/microsoft/CCF/pull/4628). This will be updated with #60 

The following changes have been made to make the constitution more testable.
- Moved the bulk of the scripts to a new folder called `scripts`
- Added make recipes for testing enclave and virtual in docker (see `make help`)
- Ensure `test.sh` executes irrespective of the hosting model (sandbox/docker)

# Todo (in separate PR)
- Move config and dockerfile into sample folder
- Fix up the path jumping that you have to do for the certs
- Find a better way to kill the python process

# Test this PR
- `make test` - this should build the app, start the sandbox, run the tests, kill the sandbox
- `make test-docker-virtual` - this will build the app, build a docker image, start the network, setup the governance and then run the tests

### CODE REVIEW

File: docker/ccf_app_js.virtual [hunk 1]
```diff
@@ -7,9 +7,8 @@ FROM mcr.microsoft.com/ccf/app/run-js:2.0.7-sgx
 # Note: libjs_generic.virtual is not included in run-js container
 COPY --from=builder /opt/ccf/lib/libjs_generic.virtual.so /usr/lib/ccf
 
-COPY --from=builder /opt/ccf/bin/*.js /app/
-
-# copy config files and member0 certificates into image
+# copy configuration into image
+COPY ./constitution/*.js /app/
```
Comment (Aymalla) [inline docker/ccf_app_js.virtual:11]: Could we read the constitution files from the run temp directory "./workspace/constitution" like certs and configs files to keep the Docker file generic and not related to each app's folder structure?

Comment (ross-p-smith) [reply docker/ccf_app_js.virtual:11]: I think going forward, each sample should have its own config, dockerfile and constitution. It could be that a particular sample needs a particular version of CCF or a different constitution. I think by keeping these things generic we are actually performing scripting gymnastics and means the docker files already are confusing as they get copied to another folder before they are built.

As an example if you look in the dockerfiles we have references to `./workspace` paths, but that folder exists in the banking-app - so it's not intuitive for people trying to work out what is going on.

I also see that we could have an IP address actually specified in the cchost_config file and each sample could have its own IP address so you could spin up the samples at the same time in docker.

In summary, I would actually like to move the dockerfile and config down a directory.

File: banking-app/test_docker.sh [hunk 2]
```diff
@@ -0,0 +1,231 @@
...
+    }
+  ]
+}
+JSON
+}
+
+##############################################
+# Discover docker configuration
+##############################################
+containerId=$(docker ps -f ancestor=banking-app:virtual -q)
```
Comment (julioalex-rezende) [inline banking-app/test_docker.sh:64]: this script will only allow you to test the virtual container image. 
is that intentional?

File: banking-app/test_docker.sh [hunk 3]
```diff
@@ -0,0 +1,231 @@
...
+echo "Network Proposal ID: $network_proposal_out_id"
+/opt/ccf/bin/scurl.sh "${server}/gov/proposals/$network_proposal_out_id/ballots" \
+    --cacert service_cert.pem \
+    --signing-key member0_privk.pem \
+    --signing-cert member0_cert.pem \
+    --data-binary @vote_accept.json \
+    -H "content-type: application/json" | jq
+
+##############################################
+# Propose users and application
```
Comment (julioalex-rezende) [inline banking-app/test_docker.sh:131]: nit:

You are proposing your users on the same step in which you open the network. not with the application

File: .devcontainer.json [hunk 4]
```diff
@@ -6,7 +6,7 @@
     "vsls-contrib.codetour",
     "esbenp.prettier-vscode"
   ],
-  "postCreateCommand": "npm install --prefix=banking-app && npm install --prefix=auditable-logging-app",
+  "postCreateCommand": "npm install --prefix=banking-app && npm install --prefix=auditable-logging-app && cp sandbox.sh /opt/ccf/bin/sandbox.sh",
```
Comment (takuro-sato) [inline .devcontainer.json:9]: Is there any issue to remove local sandbox.sh? If not, could you raise one?

File: banking-app/test_docker.sh [hunk 5]
```diff
@@ -0,0 +1,231 @@
...
+    --signing-cert member0_cert.pem \
+    --data-binary @vote_accept.json \
+    -H "content-type: application/json" | jq
+
+##############################################
+# Test Network
+##############################################
+curl "${server}/node/network" --cacert service_cert.pem | jq
+
+# -------------------------- Preparation --------------------------
```
Comment (takuro-sato) [inline banking-app/test_docker.sh:153]: Is it possible just to run/call test.sh rather than to copy the contents?

**Comment by takuro-sato** [comment]
Could you raise an issue to make the same change for auditable-logging-app?

File: .devcontainer.json (hunk 4 above)
Comment (ross-p-smith) [reply .devcontainer.json:9]: The local sandbox.sh in this repository is the same as here: -https://github.com/microsoft/CCF/commit/77fc334d87fbf3f8e13ace54a4bda391f85e3f20.

When this sandbox enters the dev container we will delete this bit of the postCreateCommand.

How do we get the latest version? We are already on mcr.microsoft.com/ccf/app/dev:3.0.0-dev6-sgx which is the latest dev container?

File: banking-app/test_docker.sh (hunk 2 above)
Comment (ross-p-smith) [reply banking-app/test_docker.sh]: Great spot - I never tested the enclave version - updated

File: banking-app/test_docker.sh (hunk 3 above)
Comment (ross-p-smith) [reply banking-app/test_docker.sh:175]: Yes, this is because we use the set_app_js that is generated when we build the actual app - I have added a new story so that we can make this all a bit better: - #58

File: .devcontainer.json (hunk 4 above)
Comment (takuro-sato) [reply .devcontainer.json:9]: Sorry after looking at my comment again, I found it not clear.
I meant that I want to have a github issue not to forget to delete it.

We can only use the Sandbox of https://github.com/microsoft/CCF/pull/4628 only after it's included in a release.

Comment (ross-p-smith) [reply .devcontainer.json:9]: Captured in #60

File: banking-app/test_docker.sh (hunk 5 above)
Comment (ross-p-smith) [reply banking-app/test_docker.sh]: I've just reworked the whole PR - I hope this is better - let me know your thoughts

**Comment by ross-p-smith** [comment]
> Could you raise an issue to make the same change for auditable-logging-app?

#57

File: banking-app/Makefile [hunk 6]
```diff
@@ -10,23 +10,45 @@ build: ## 🔨 Build the Banking Application
 	@echo -e "\e[34m$@\e[0m" || true
 	@npm run build
 
-test: ## 🧪 Test the Banking Application
+test: start-host-for-test ## 🧪 Test the Banking Application in the sandbox
```
Comment (takuro-sato) [inline banking-app/Makefile:13]: How about creating test_local.sh instead of `start-host-for-test` and `stop_sandbox.sh`?
test_local.sh would be something like this (it's very rough, but hopefully you will get the idea):
```bash
<ccf>/sandbox.sh <options>
sandbox_pid=$!

function finish {
  kill -9 $sandbox_pid
}
trap finish EXIT

# waiting for sandbox etc...

./test.sh
```

It would be safer in terms of stopping correct process (sandbox).

File: banking-app/scripts/start_docker.sh [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+fi
+if [ -z $serverIP ]; then
+    failed "You must supply --serverIP"
+fi
+
+docker run --detach --ip $serverIP banking-app:$enclave_type
+containerId=$(docker ps -f ancestor=banking-app:$enclave_type -q)
+# Wait for container to generate the cert
+echo "Waiting for docker to startup . . ."
+sleep 5
```
Comment (takuro-sato) [inline banking-app/scripts/start_docker.sh:57]: Considering that flaky CI caused by sleep is hard to find the reason, is it worth to write a code to actually wait for startup rather than to sleep for a fixed amount of time?

I imagine something similer to test.sh
https://github.com/microsoft/ccf-app-samples/blob/f3cbe4234252c619ca294c17c0bf7ab74bcc331f/banking-app/scripts/test.sh#L67

File: banking-app/scripts/setup_governance.sh [hunk 8]
```diff
@@ -0,0 +1,181 @@
...
+    exit 1
+}
+
+# parse parameters
+
+if [[ $# -lt 2 || $# -gt 2 ]]; then
+    usage
+    exit 1
+fi
+
```
Comment (julioalex-rezende) [inline banking-app/scripts/setup_governance.sh:28]: I haven't tested this, but looks like the `--help` switch (which is used alone) will not work if you have this validation.
The final result will be the same, as `--help` outputs the usage function, so in the way it's written, the output is correct, however not through the correct flow. If I'm not mistaken we may be adding a trap here :)

You can leave the checks to the for loop code below, and the correct parameter check will be made on lines 41-45

File: config/cchost_config_virtual_js.json [hunk 9]
```diff
@@ -1,32 +1,34 @@
...
-    "enclave": {
-      "file": "/usr/lib/ccf/libjs_generic.virtual.so",
-      "type": "Virtual"
+  "enclave": {
+    "file": "/usr/lib/ccf/libjs_generic.virtual.so",
+    "type": "Virtual"
+  },
+  "network": {
+    "node_to_node_interface": {
+      "bind_address": "172.17.0.3:8081"
```
Comment (julioalex-rezende) [inline config/cchost_config_virtual_js.json:8]: Should we make changes to the enclave config file as well? (ok if done separately, just raising the question)

**Review by julioalex-rezende — CHANGES_REQUESTED** [review]
Open to discuss any of the raised comments for a better alignment if necessary.

File: banking-app/scripts/setup_governance.sh (hunk 8 above)
Comment (ross-p-smith) [reply banking-app/scripts/setup_governance.sh]: Great spot - I have changed this to always check greater than the maximum number of parameters so that help works.  If you are happy, can you resolve.

File: banking-app/scripts/start_docker.sh [hunk 10]
```diff
@@ -0,0 +1,58 @@
...
+    printf "Script failed: %s\n\n" "$1"
+    exit 1
+}
+
+# parse parameters
+
+if [[ $# -lt 3 || $# -gt 3 ]]; then
+    usage
+    exit 1
+fi
```
[the remark this answers is not in the document]
Comment (ross-p-smith) [reply banking-app/scripts/start_docker.sh]: Great spot - I have changed this to always check greater than the maximum number of parameters so that help works.  If you are happy, can you resolve.

File: banking-app/Makefile (hunk 6 above)
Comment (ross-p-smith) [reply banking-app/Makefile]: Great plan - what do you think now @takuro-sato?

File: banking-app/scripts/start_docker.sh (hunk 7 above)
Comment (ross-p-smith) [reply banking-app/scripts/start_docker.sh]: Agreed - done!

File: banking-app/Makefile [hunk 11]
```diff
@@ -10,23 +10,28 @@ build: ## 🔨 Build the Banking Application
...
 # Run sandbox. Consider 3 members as 3 banks.
-start-host: build ## 🏃 Start the CCF Sandbox
+# This is used in the demo scripts
+start-host: build ## 🏁 Start the CCF Sandbox for the demo
 	@echo -e "\e[34m$@\e[0m" || true
-	@cp constitutions/resolve.js /opt/ccf/bin/resolve.js
-	@/opt/ccf/bin/sandbox.sh --js-app-bundle ./dist/ --initial-member-count 3 --initial-user-count 0
+	@/opt/ccf/bin/sandbox.sh --js-app-bundle ./dist/ --initial-member-count 3 --initial-user-count 0 --constitution-dir ./constitution
 
 demo: ## 🎬 Demo the Banking Application
```
Comment (Aymalla) [inline banking-app/Makefile:31]: `make demo` throw exception "Failed to connect to 127.0.0.1 port 8000: Connection refused"

**Review by Aymalla — APPROVED** [review]
LGTM, Just one comment

File: banking-app/Makefile (hunk 11 above)
Comment (ross-p-smith) [reply banking-app/Makefile:31]: This is because when you "demo" you need to run `make start-host` in another terminal. I have added a check to see if the sandbox is running and tell the user.

make test - creates and tears down each time
make demo - is an interactive session with a network.

File: banking-app/scripts/setup_governance.sh [hunk 12]
```diff
@@ -0,0 +1,181 @@
...
+    --data-binary @../../dist/set_js_app.json \
+    -H "content-type: application/json")
+application_proposal_out_id=$( jq -r  '.proposal_id' <<< "${application_proposal_out}" )
+echo "Application Proposal ID: ${application_proposal_out_id}"
+/opt/ccf/bin/scurl.sh "${server}/gov/proposals/${application_proposal_out_id}/ballots" \
+    --cacert service_cert.pem \
+    --signing-key member0_privk.pem \
+    --signing-cert member0_cert.pem \
+    --data-binary @vote_accept.json \
+    -H "content-type: application/json" | jq
```
[the remark this answers is not in the document]
Comment (ross-p-smith) [reply banking-app/scripts/setup_governance.sh]: I've added a comment on where the set_js_app.json file comes from.

File: config/cchost_config_virtual_js.json (hunk 9 above)
Comment (ross-p-smith) [reply config/cchost_config_virtual_js.json:8]: I'd rather tidy up these config files as part of #56

**Review by julioalex-rezende — APPROVED** [review]
